File: go/internal/commands/events.go
package commands

import (
	"encoding/json"
	"fmt"
	"os"
	"strconv"
	"strings"

	"all-you-need-is-git/go/internal/gitx"
)

type EventsOptions struct {
	History bool
	Limit   int
	JSON    bool
}

type Event struct {
	Commit   string              `json:"commit"`
	Date     string              `json:"date"`
	Message  string              `json:"message"`
	State    *string             `json:"state"`
	RunID    *string             `json:"runId"`
	Origin   *string             `json:"originState"`
	Trailers map[string][]string `json:"trailers"`
}

func Events(opts EventsOptions) error {
	if _, err := gitx.Run("", "rev-parse", "--show-toplevel"); err != nil {
		return fmt.Errorf("Error: Not a Git repository. Please run `git init` first.")
	}

	limit := 1
	if opts.History {
		limit = opts.Limit
		if limit < 1 {
			limit = 1
		}
	}

	raw, err := gitx.Run("", "log", "-n", strconv.Itoa(limit), "--format=%H%x1f%cI%x1f%B%x1e")
	if err != nil {
		return err
	}

	records := strings.Split(raw, "\x1e")
	events := make([]Event, 0)
	for _, record := range records {
		if strings.TrimSpace(record) == "" {
			continue
		}
		parts := strings.Split(record, "\x1f")
		commit := ""
		date := ""
		message := ""
		if len(parts) > 0 {
			commit = parts[0]
		}
		if len(parts) > 1 {
			date = parts[1]
		}
		if len(parts) > 2 {
			message = strings.Join(parts[2:], "\x1f")
		}

		firstLine, body := splitCommitMessage(message)
		trailers, parseErr := parseTrailersFromBody(body)
		if parseErr != nil {
			return parseErr
		}
		stateValue := trailerValue(trailers, "aynig-state")
		runIDValue := trailerValue(trailers, "aynig-run-id")
		originValue := trailerValue(trailers, "aynig-origin-state")

		var statePtr *string
		if stateValue != "" {
			statePtr = &stateValue
		}
		var runIDPtr *string
		if runIDValue != "" {
			runIDPtr = &runIDValue
		}
		var originPtr *string
		if originValue != "" {
			originPtr = &originValue
		}

		events = append(events, Event{
			Commit:   commit,
			Date:     date,
			Message:  firstLine,
			State:    statePtr,
			RunID:    runIDPtr,
			Origin:   originPtr,
			Trailers: trailers,
		})
	}

	if opts.JSON {
		encoder := json.NewEncoder(os.Stdout)
		encoder.SetIndent("", "  ")
		return encoder.Encode(events)
	}

	if len(events) == 0 {
		fmt.Println("No commits found.")
		return nil
	}

	for _, event := range events {
		shortCommit := "unknown"
		if event.Commit != "" {
			if len(event.Commit) > 7 {
				shortCommit = event.Commit[:7]
			} else {
				shortCommit = event.Commit
			}
		}
		state := "n/a"
		if event.State != nil {
			state = *event.State
		}
		runID := "n/a"
		if event.RunID != nil {
			runID = *event.RunID
		}
		origin := ""
		if event.Origin != nil {
			origin = " origin=" + *event.Origin
		}
		fmt.Printf("%s %s state=%s run=%s%s %s\n", shortCommit, event.Date, state, runID, origin, event.Message)
	}

	return nil
}

File: js/commands/events.js
import { git } from '../gitHelpers/git.js';
import { parseCommitMessage } from '../gitHelpers/parseCommitMessage.js';

function splitCommitMessage(fullMessage) {
  const normalized = fullMessage.replace(/\r\n/g, '\n');
  const lines = normalized.split('\n');
  const firstLine = lines.shift() || '';
  let body = lines.join('\n');
  body = body.replace(/^\n+/, '');
  return { firstLine, body };
}

function normalizeTrailerValue(value) {
  if (Array.isArray(value)) {
    return value[value.length - 1];
  }
  return value;
}

function toLowerCaseKeys(trailers) {
  const out = {};
  for (const [key, value] of Object.entries(trailers || {})) {
    out[key.toLowerCase()] = value;
  }
  return out;
}

async function action(options) {
  try {
    const historyEnabled = Boolean(options.history);
    let limit = historyEnabled ? Number.parseInt(options.limit, 10) : 1;
    if (!Number.isFinite(limit) || limit < 1) {
      limit = 1;
    }

    const raw = await git.raw([
      'log',
      '-n',
      String(limit),
      '--format=%H%x1f%cI%x1f%B%x1e'
    ]);

    const records = raw.split('\x1e').filter(Boolean);
    const events = [];

    for (const record of records) {
      const parts = record.split('\x1f');
      const commit = parts[0] || '';
      const date = parts[1] || '';
      const message = parts.slice(2).join('\x1f');
      const { firstLine, body } = splitCommitMessage(message);
      const { trailers } = await parseCommitMessage({ message: firstLine, body });
      const trailersLower = toLowerCaseKeys(trailers);
      const state = normalizeTrailerValue(trailersLower['aynig-state']);
      const runId = normalizeTrailerValue(trailersLower['aynig-run-id']);
      const originState = normalizeTrailerValue(trailersLower['aynig-origin-state']);

      events.push({
        commit,
        date,
        message: firstLine,
        state: state || null,
        runId: runId || null,
        originState: originState || null,
        trailers
      });
    }

    if (options.json) {
      console.log(JSON.stringify(events, null, 2));
      return;
    }

    if (events.length === 0) {
      console.log('No commits found.');
      return;
    }

    for (const event of events) {
      const shortCommit = event.commit ? event.commit.slice(0, 7) : 'unknown';
      const state = event.state || 'n/a';
      const runId = event.runId || 'n/a';
      const origin = event.originState ? ` origin=${event.originState}` : '';
      console.log(`${shortCommit} ${event.date} state=${state} run=${runId}${origin} ${event.message}`);
    }
  } catch (error) {
    console.error('Error: Not a Git repository. Please run `git init` first.');
    process.exit(1);
  }
}

export function registerEventsCommand(program) {
  program
    .command('events')
    .description('Show recent AYNIG events from commit trailers')
    .option('--history', 'Allow scanning recent history')
    .option('-n, --limit <number>', 'Number of commits to inspect (requires --history)', '10')
    .option('--json', 'Output JSON')
    .action(action);
}
